### PR TITLE
Issue #204 - Support (Oracle) missing types added

### PR DESCRIPTION
**Pull request by ancientwizard**
- SQLT_TIMESTAMP(TIMESTAMP)
- SQLT_TIMESTAMP_LTZ(TIMESTAMP WITH LOCAL TIME ZONE)

As requested one issue, one change set. This should be small and simple enough to pull. The only change in behavior is DBIish will no longer complain and throw exception for about two missing types of TIMESTAMP.

### CODE REVIEW

**Comment by ancientwizard** [comment]
BTW: I'm successfully running the first of three unit test's that validate Oracle database operations against applications schema.

**Review by jonathanstowe — APPROVED** [review]

**Review by jonathanstowe — APPROVED** [review]

**Comment by jonathanstowe** [comment]
I *think* this is a no-brainer but I'd be happier if someone else gave it a :+1:  before I merge.

**Comment by rbt** [comment]
Should a test or 2 be added to t/57-oracle-datetime.t to ensure the various timezone differences are handled as expected?

Otherwise, it appears good to go.

**Comment by ancientwizard** [comment]
I'll have a look at the test set and see about adding tests. Thanks for the motivation; I hadn't tried to get them to work s I suspect they had lots of dependencies. I'll figure something out

**Comment by ancientwizard** [comment]
Tests have been added. I did run issue trying to get prove to run. Best guess it was using perl5 to run the test because the file are "*.t"; it might be time to rename them as .t6? I cobbled it together using this command. It should be good enough.

[vburns@localhost DBIish]$ LD_LIBRARY_PATH=$ORACLE_HOME/lib PERL6LIB=./lib DBIISH_WRITE_TEST=YES perl6 t/57-oracle-datetime.t
1..33
ok 1 - Connected
ok 2 - Clean
ok 3 - Table created
ok 4 - Can insert Perl6 values
ok 5 - Column-types match
ok 6 - The object is-a 'Date'
ok 7 - The object is-a 'DateTime'
ok 8 - Today
ok 9 - Right now
ok 10 - SYSDATE is Date
ok 11 - Today
ok 12 - CURRENT_TIMESTAMP is DateTime
ok 13 - Server drift
# 0.224323
ok 14 - Delete rows
ok 15 - Close session
#+ -------------------------------------------------------- +
## Issue #204 - ADDED Missing TIMESTAMP Support (BEGIN)
#+ -------------------------------------------------------- +
ok 16 - ALTER SESSION ...
ok 17 - ALTER SESSION ...
ok 18 - ALTER SESSION ...
ok 19 - ALTER SESSION ...
ok 20 - Perfect table is empty!
# Tasty timestamp: 2021-05-29T02:28:06Z
ok 21 - Can insert Perl6 values
ok 22 - One row!
ok 23 - DATE is read into Date.new() but lacks the time component!
ok 24 - TIMESTAMP(0) is read into DateTime.new()
ok 25 - TIMESTAMP(6) is read into DateTime.new()
ok 26 - TIMESTAMP(0) W TZ is read into DateTime.new()
ok 27 - TIMESTAMP(6) W TZ is read into DateTime.new()
ok 28 - TIMESTAMP(0) W TZ is read into DateTime.new()
ok 29 - TIMESTAMP(6) W TZ is read into DateTime.new()
ok 30 - dispose
ok 31 - Perfect table is empty!
#+ -------------------------------------------------------- +
## Issue #204 - ADDED Missing TIMESTAMP Support (END)
#+ -------------------------------------------------------- +
ok 32 - Table Cleanup
ok 33 - Say good night!

**Comment by jonathanstowe** [comment]
@salortiz @abraxxa as the other people who have touched the Oracle driver, any other input?

**Review by salortiz — APPROVED** [review]

File: t/57-oracle-datetime.t [hunk 1]
```diff
@@ -30,7 +30,8 @@ without $dbh {
 ok $dbh,    'Connected';
 my $dropper = q|
     BEGIN
-       EXECUTE IMMEDIATE 'DROP TABLE test_datetime';
+       EXECUTE IMMEDIATE 'PURGE RECYCLEBIN';
```
Comment (abraxxa) [inline t/57-oracle-datetime.t:33]: Why did you add this statement? In a shared test database it might have unwanted side effects.

File: t/57-oracle-datetime.t [hunk 2]
```diff
@@ -83,6 +89,73 @@ isa-ok $sth.column-types[0], DateTime, 'CURRENT_TIMESTAMP is DateTime';
...
 isnt $datetime, $datetime2,		    'Server drift';
+
 diag $datetime2.Instant - $datetime.Instant;
 
-$dbh.execute($dropper);
+ok $dbh.execute('DELETE FROM test_datetime WHERE adate IS NOT NULL'), 'Delete rows';
+ok $dbh.dispose, 'Close session';
+
+note '#+ -------------------------------------------------------- +';
+note '## Issue #204 - ADDED Missing TIMESTAMP Support (BEGIN)';
```
Comment (abraxxa) [inline t/57-oracle-datetime.t:99]: I prefer 'add' to 'added' according to the git-book regarding commit messages.

File: t/57-oracle-datetime.t [hunk 3]
```diff
@@ -83,6 +89,73 @@ isa-ok $sth.column-types[0], DateTime, 'CURRENT_TIMESTAMP is DateTime';
...
+    VALUES (?,?,?,?,?,?,?)|);
+$now = DateTime.now.truncated-to('second').in-timezone(0);
+
+say '# Tasty timestamp: ', $now;
+lives-ok {
+    $sth.execute(
+        $now, # Need an ISO full date
+	$now, $now, $now, $now, $now, $now
+    );
+},                                           'Can insert Perl6 values';
```
Comment (abraxxa) [inline t/57-oracle-datetime.t:133]: This should read Raku not Perl6.

File: t/57-oracle-datetime.t [hunk 4]
```diff
@@ -83,6 +89,73 @@ isa-ok $sth.column-types[0], DateTime, 'CURRENT_TIMESTAMP is DateTime';
...
+is $dbh.execute('SELECT atimestamp6 FROM test_datetime').allrows, $now.Str, 'TIMESTAMP(6) is read into DateTime.new()';
+# TIMESTAMP(0)/(N) WITH TIME ZONE
+is $dbh.execute('SELECT atimestamp0tz FROM test_datetime').allrows, $now.Str, 'TIMESTAMP(0) W TZ is read into DateTime.new()';
+is $dbh.execute('SELECT atimestamp6tz FROM test_datetime').allrows, $now.Str, 'TIMESTAMP(6) W TZ is read into DateTime.new()';
+# TIMESTAMP(0)/(N) WITH LOCAL TIME ZONE
+is $dbh.execute('SELECT atimestamp0ltz FROM test_datetime').allrows, $now.Str, 'TIMESTAMP(0) W LTZ is read into DateTime.new()';
+is $dbh.execute('SELECT atimestamp6ltz FROM test_datetime').allrows, $now.Str, 'TIMESTAMP(6) W LTZ is read into DateTime.new()';
+
+$dbh.rollback;
+ok $sth.dispose,  'dispose';
```
Comment (abraxxa) [inline t/57-oracle-datetime.t:150]: I don't think we need to test the return value of dispose.

File: t/57-oracle-datetime.t [hunk 5]
```diff
@@ -83,6 +89,73 @@ isa-ok $sth.column-types[0], DateTime, 'CURRENT_TIMESTAMP is DateTime';
...
+is $dbh.execute('SELECT * FROM test_datetime').allrows.elems, 0, 'Perfect table is empty!';
+
+note '#+ -------------------------------------------------------- +';
+note '## Issue #204 - ADDED Missing TIMESTAMP Support (END)';
```
Comment (abraxxa) [inline t/57-oracle-datetime.t:154]: Same as above.

File: t/57-oracle-datetime.t [hunk 6]
```diff
@@ -83,6 +89,73 @@ isa-ok $sth.column-types[0], DateTime, 'CURRENT_TIMESTAMP is DateTime';
...
+note '#+ -------------------------------------------------------- +';
+
+# Done
+ok $dbh.execute($dropper), 'Table Cleanup';
+ok $dbh.dispose, 'Say good night!';
```
Comment (abraxxa) [inline t/57-oracle-datetime.t:159]: Also as above, those don't need to be tests. Might be better to move them to an END block (or something similar to a Perl 5 END block).

**Review by abraxxa — CHANGES_REQUESTED** [review]

File: t/57-oracle-datetime.t (hunk 1 above)
Comment (ancientwizard) [reply t/57-oracle-datetime.t:33]: It's called cleaning up after yourself; This test left unchecked (I'm sure other test scripts are dirty as well) will waste resources every time a table is created and dropped. Its my experience that folks will go unaware that the RECYCELEBIN is filling up or that it even exists. It will produce no negative issues. Anyone relying on the RECYCLEBIN as a long time storage recovery of discarded tables is looking for trouble. We're doing folks a favor.

File: t/57-oracle-datetime.t (hunk 2 above)
Comment (ancientwizard) [reply t/57-oracle-datetime.t:99]: feel free to change to comment all you like; Seem hardly worth the effort.

File: t/57-oracle-datetime.t (hunk 3 above)
Comment (ancientwizard) [reply t/57-oracle-datetime.t:133]: I agree; but I didn't write the comment; I.E. you'll find the same elsewhere in the test; I followed the example; since I'd already gotten grief for too many changes

File: t/57-oracle-datetime.t (hunk 6 above)
Comment (ancientwizard) [reply t/57-oracle-datetime.t:159]: No such thing as too many tests; Feel free to do whatever with them.

File: t/57-oracle-datetime.t (hunk 2 above)
Comment (jonathanstowe) [reply t/57-oracle-datetime.t:99]: Frankly I'd completely lose the note in the file, it serves no useful purpose - most people won't see the message and anyone who is interested can see the commit message,

Comment (ancientwizard) [reply t/57-oracle-datetime.t:99]: notes like this serve  two purposes
- It acts like a script comment
- It demarcates sections of a test; the next poor sap making changes will have yet another clue about what is taking place

I was tempted to write it as a subtest which would do something similar; but I didn't see it being used

File: t/57-oracle-datetime.t (hunk 3 above)
Comment (abraxxa) [reply t/57-oracle-datetime.t]: Line 133 is displayed as added by your commit in github.

**Comment by ancientwizard** [comment]
Old comment was nuked from orbit. The bad old aliens are gone. It's safe to come out now. Really I promise! Just don't look over there under your bed, that last egg thing is being saved for you.

ADD vs. ADDED = /rəˈdikyələs/

File: t/57-oracle-datetime.t (hunk 3 above)
Comment (ancientwizard) [reply t/57-oracle-datetime.t]: I now says Raku and the comment was nuked

**Comment by ancientwizard** [comment]
Taking silence to equal "not-interested" - I resend my help aka pull request.
If you change your mind I have a progression of branches issue-204, issue-203 & issue-207 that build on each other; all are merged into master for the complete solution. - Cheers!
